akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringEvents.scala

```diff
@@ -0,0 +1,185 @@
+    Behaviors.setup { _ ⇒
+      startRecoveryTimer(setup.timers, setup.settings.recoveryEventTimeout)
+
+      replayEvents(state.seqNr + 1L, setup.recovery.toSequenceNr)
+
+      withMdc {
+        stay(state)
+      }
+    }
+
+  }
+
+  private def stay(state: RecoveringState[S]): Behavior[InternalProtocol] =
+    Behaviors.immutable {
+      case (_, JournalResponse(r))       ⇒ onJournalResponse(state, r)
+      case (_, SnapshotterResponse(r))   ⇒ onSnapshotterResponse(r)
+      case (_, RecoveryTickEvent(snap))  ⇒ onRecoveryTick(state, snap)
+      case (_, cmd @ IncomingCommand(_)) ⇒ onCommand(cmd)
+    }
+
+  private def withMdc(wrapped: Behavior[InternalProtocol]) = {
+    val mdc = Map(
+      "persistenceId" → setup.persistenceId,
+      "phase" → "recover-evnts"
+    )
+
+    Behaviors.withMdc((_: Any) ⇒ mdc, wrapped)
+  }
+
+  private def onJournalResponse(
+    state:    RecoveringState[S],
+    response: JournalProtocol.Response): Behavior[InternalProtocol] = {
+    import setup.context.log
+    try {
+      response match {
+        case ReplayedMessage(repr) ⇒
+          // eventSeenInInterval = true
+          // updateLastSequenceNr(repr)
+
+          val newState = state.copy(
+            seqNr = repr.sequenceNr,
+            state = setup.eventHandler(state.state, repr.payload.asInstanceOf[E])
+          )
+
+          stay(newState)
+
+        case RecoverySuccess(highestSeqNr) ⇒
+          log.debug("Recovery successful, recovered until sequenceNr: {}", highestSeqNr)
+          cancelRecoveryTimer(setup.timers)
+
+          try onRecoveryCompleted(state)
+          catch { case NonFatal(ex) ⇒ onRecoveryFailure(ex, highestSeqNr, Some(state)) }
+
+        case ReplayMessagesFailure(cause) ⇒
+          onRecoveryFailure(cause, state.seqNr, None)
+
+        case other ⇒
+          //          stash(setup, setup.internalStash, other)
+          //          Behaviors.same
+          Behaviors.unhandled
+      }
+    } catch {
+      case NonFatal(cause) ⇒
+        cancelRecoveryTimer(setup.timers)
+        onRecoveryFailure(cause, state.seqNr, None)
+    }
+  }
+
+  private def onCommand(cmd: InternalProtocol): Behavior[InternalProtocol] = {
+    // during recovery, stash all incoming commands
+    stash(cmd)
+    Behaviors.same
+  }
+
+  protected def onRecoveryTick(state: RecoveringState[S], snapshot: Boolean): Behavior[InternalProtocol] =
+    if (!snapshot) {
+      if (state.eventSeenInInterval) {
+        stay(state.copy(eventSeenInInterval = false))
+      } else {
+        cancelRecoveryTimer(setup.timers)
+        val msg = s"Recovery timed out, didn't get event within ${setup.settings.recoveryEventTimeout}, highest sequence number seen ${state.seqNr}"
+        onRecoveryFailure(new RecoveryTimedOut(msg), state.seqNr, None) // TODO allow users to hook into this?
+      }
+    } else {
+      // snapshot timeout, but we're already in the events recovery phase
+      Behavior.unhandled
```

Comment:
That's the nice thing with timers, once you canceled it will not be received even though it's already enqueued in mailbox. It's ok to keep the unhandled for this. I just wanted to point out that it shouldn't happen.

Reply:
Ah excellent, I had forgotten about this :)